### ISSUE TITLE
Add typings

### ISSUE THREAD

**Issue by bbobrian** [Breeze/breeze.js.labs#29]
It would be great to have better TypeScript support.

**Comment by marcelgood** [comment]
In what specific context are you looking for TypeScript support? Most of the labs stuff extends existing breeze types by adding methods to existing types. or transparently hooking into breeze under the hood. They don't directly offer new types.

With TypeScript 1.8 we can now define augmentation modules, but that requires that you are using external TypeScript modules. This would be the only way to provide typings.

I just pushed a fix, so that these extensions load correctly as an external module. With this fix you can do this now TypeScript for example.

`import 'breeze-client-labs/breeze.ajaxpost';`

This will load breeze.ajaxpost.js and patch breeze, but it requires that you load breeze as an external module. We have updated the type definitions for the breeze-client npm package recently to support the loading of breeze as an external module. To learn more take a look at our new Angular 2 bridge, where this comes in handy.

[Breeze-Angular2 bridge](https://github.com/Breeze/breeze.bridge.angular2/blob/master/README.md)

We can look into adding augmentation module definitions to the repo and npm package, but you would only benefit from that if your application uses external TypeScript modules. If you are statically loading breeze and the breeze.labs extension, then you wouldn't get the typing support.

**Comment by bbobrian** [comment]
Yes that is exactly what I mean. I am importing breeze as an external module, and I also want to use the MetadataHelper. Currently I import both as follows:

```typescript
import * as breeze from 'breeze-client';
import 'breeze-client-labs/breeze.metadata-helper';
```

I then have full TypeScript support for the core breeze client properties/methods, but if I try to use `breeze.config.MetadataHelper` TypeScript will complain that it doesn't exist.

I've created my own metadataHelper.d.ts with the methods I need to use:

```typescript
import * as breeze from 'breeze-client';

export class MetadataHelper {
    constructor(modelName: string, keyTypeEnum: any);

    addDataService(store: breeze.MetadataStore, serviceName: string): void;
    addTypeToStore(store: breeze.MetadataStore, typeDef: any): breeze.IStructuralType;
}
```

And then I can do something ugly to use it:

```typescript
private helper = <MetadataHelper>(<any>breeze.config).MetadataHelper('myModel', breeze.AutoGeneratedKeyType.KeyGenerator);
```

This is my first project using TypeScript, so I hadn't heard of module augmentation. I will look into that to make my implementation cleaner until the full type definitions are available for labs.

**Comment by marcelgood** [comment]
I've pushed type definitions for the MetadataHelper to the repo. 

https://github.com/Breeze/breeze.js.labs/blob/master/breeze.metadata-helper.d.ts

For now, copy this file to node_modules/breeze-client-labs. That should give you full TypeScript support for the MetadataHelper. Let me know if it works for you. I then look into adding type definitions for the other extensions where applicable and publish an updated npm package.

**Comment by bbobrian** [comment]
Working great! Thanks!

**Comment by marcelgood** [comment]
I've published the breeze-client-labs@1.5.11 npm package containing the TypeScript support. If you upgrade to the latest package you should get all the type definitions now for most lab extensions. There were two or three that I couldn't provide type information for, because they extend the augment the top level scope, which is not supported by TypeScript. We'll have to rewrite those extensions first, but that's for another day maybe.
